<a href="https://colab.research.google.com/github/guebin/MP2024/blob/main/quiz/Quiz-4.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" style="text-align: left"></a>

| **항목**               | **허용 여부**        | **비고**                                          |
|------------------------|----------------------|---------------------------------------------------|
| **강의노트 참고**      | 허용                 | 수업 중 제공된 강의노트나 본인이 정리한 자료를 참고 가능       |
| **구글 검색**          | 허용                 | 인터넷을 통한 자료 검색 및 정보 확인 가능        |
| **생성 모형 사용**           | 허용 안함            | 인공지능 기반 도구(GPT 등) 사용 불가            |

In [1]:
import datasets
import transformers
import torchvision.transforms
import evaluate
import numpy as np

/home/cgb3/anaconda3/envs/hf/lib/python3.12/site-packages/tqdm/auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm


아래는 `datasets` 라이브러리의 `load_dataset` 함수를 사용하여 `beans`라는 데이터셋을 불러오는 코드이다. 

In [6]:
beans = datasets.load_dataset('beans')
beans

DatasetDict({
    train: Dataset({
        features: ['image_file_path', 'image', 'labels'],
        num_rows: 1034
    })
    validation: Dataset({
        features: ['image_file_path', 'image', 'labels'],
        num_rows: 133
    })
    test: Dataset({
        features: ['image_file_path', 'image', 'labels'],
        num_rows: 128
    })
})

`beans` 데이터셋은 **train**(1034개), **validation**(133개), **test**(128개)로 나뉘며, 각각 모델 학습, 검증, 성능 평가에 사용된다. 각 데이터셋은 이미지 파일 경로(`image_file_path`), 실제 이미지 데이터(`image`), 그리고 이미지에 대한 질병 라벨(`labels`)로 구성되어 있다. 이 라벨은 각 이미지가 어떤 질병을 나타내는지에 대한 정보를 제공한다.

라벨 0은 "각진 잎 반점(angular_leaf_spot)"이라는 질병으로, 이는 콩 잎에 특정 형태의 반점이 생기는 현상을 의미한다. 라벨 1은 "콩 녹병(bean_rust)"으로, 콩 잎에 붉은 녹병이 나타나는 질병이다. 마지막으로 라벨 2는 "healthy(건강한 잎)"을 나타내며, 이 라벨이 붙은 이미지는 질병 없이 건강한 콩 잎을 뜻한다. 이러한 라벨을 통해 데이터셋의 각 이미지가 어떤 질병을 가지고 있는지 쉽게 파악할 수 있다.

In [7]:
id2label = {0: 'angular_leaf_spot', 1: 'bean_rust', 2: 'healthy'}
label2id = {'angular_leaf_spot': 0, 'bean_rust': 1, 'healthy': 2}

# 1. `beans` 자료분석 -- 100점

> 배점: (1),(2),(3),(4) -- 5점, (5),(6) -- 30점 (7),(8) -- 10점